mne/preprocessing/xdawn.py

```diff
@@ -0,0 +1,565 @@
+        exclude : array_like of int.
+            The indices refering to columns in the ummixing matrix. The
+            components to be zeroed out.
+
+        Returns
+        -------
+        out : dict of instance
+            A dict of instance (from the same type as inst input) for each
+            event type in event_id.
+        """
+        if event_id is None:
+            event_id = self.event_id
+
+        if isinstance(inst, _BaseRaw):
+            out = self._apply_raw(raw=inst, include=include,
+                                  exclude=exclude, event_id=event_id)
+        elif isinstance(inst, _BaseEpochs):
+            out = self._apply_epochs(epochs=inst, include=include,
+                                     exclude=exclude, event_id=event_id)
+        elif isinstance(inst, Evoked):
+            out = self._apply_evoked(evoked=inst, include=include,
+                                     exclude=exclude, event_id=event_id)
+        else:
+            raise ValueError('Data input must be of Raw, Epochs or Evoked '
+                             'type')
+        return out
+
+    def _apply_raw(self, raw, include, exclude, event_id):
+        """Aux method."""
+        if not raw.preload:
+            raise ValueError('Raw data must be preloaded to apply Xdawn')
+
+        if exclude is None:
+            exclude = list(set(self.exclude))
+        else:
+            exclude = list(set(self.exclude + exclude))
+
+        picks = pick_types(raw.info, meg=False, include=self.ch_names,
+                           exclude='bads')
+        raws = dict()
+        for eid in event_id:
+            data = raw[picks, :][0]
+
+            data = self._pick_sources(data, include, exclude, eid)
+
+            raw_r = raw.copy()
+
+            raw_r[picks, :] = data
+            raws[eid] = raw_r
+        return raws
+
+    def _apply_epochs(self, epochs, include, exclude, event_id):
+        """Aux method."""
+        if not epochs.preload:
+            raise ValueError('Epochs must be preloaded to apply Xdawn')
+
+        picks = pick_types(epochs.info, meg=False, ref_meg=False,
+                           include=self.ch_names,
+                           exclude='bads')
+
+        # special case where epochs come picked but fit was 'unpicked'.
+        if len(picks) != len(self.ch_names):
+            raise RuntimeError('Epochs don\'t match fitted data: %i channels '
+                               'fitted but %i channels supplied. \nPlease '
+                               'provide Epochs compatible with '
+                               'xdawn.ch_names' % (len(self.ch_names),
+                                                   len(picks)))
+
+        epochs_dict = dict()
+        data = np.hstack(epochs.get_data()[:, picks])
+
+        for eid in event_id:
+
+            data_r = self._pick_sources(data, include, exclude, eid)
+
+            epochs_r = epochs.copy()
+            epochs_r._data[:, picks] = np.array(np.split(data_r,
```

Comment:
Do you have a good way to affect data to an existing Epoch instance ?
Note : this is actual code from ICA